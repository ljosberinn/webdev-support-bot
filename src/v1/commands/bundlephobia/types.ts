export interface BundlephobiaResponse {
  package: {
    name: string;
    scope: string;
    version: string;
    description: string;
    keywords: string[];
    date: Date;
    links: {
      npm: string;
      homepage: string;
      repository: string;
      bugs: string;
    };
    publisher: {
      username: string;
      email: string;
    };
    maintainers: {
      username: string;
      email: string;
    }[];
    author: {
      name: string;
      email: string;
      url: string;
      username: string;
    };
  };
  score: {
    final: number;
    detail: {
      quality: number;
      popularity: number;
      maintenance: number;
    };
  };
  searchScore: number;
  highlight: string;
}

export interface ExtendedBundlephobiaResponse {
  assets: { gzip: number; name: string; size: string; type: string }[];
  dependencyCount: number;
  gzip: number;
  size: number;
  scoped: boolean;
  hasSideEffects: boolean;
  hasJSNext: boolean;
  hasJSModule: boolean;
  version: string;
  name: string;
  repository: string;
  description: string;
  dependencySizes: { approximateSize: number; name: string }[];
}

export interface Fields {
  gzip: number;
  previousVersionSize?: number;
  dependencies: number;
  hasSideEffects: boolean;
  isTreeShakeable: boolean;
  estDownloadTimeEdge: string;
  estDownloadTimeEmerging3g: string;
  similarPackages: {
    label: string | undefined;
    packages: { name: string; size: number }[];
  };
}

export interface SimilarPackagesResponse {
  name: string;
  category: {
    label: string;
    score: number;
    tags: { tag: string; weight: number }[];
    similar: string[];
  };
}
