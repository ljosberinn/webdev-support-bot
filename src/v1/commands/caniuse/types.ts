export interface Ie {
  [key: string]: string;
}

export interface Edge {
  [key: string]: string;
}

export interface Firefox {
  [key: string]: string;
}

export interface Chrome {
  [key: string]: string;
}

export interface Safari {
  [key: string]: string;
}

export interface Opera {
  [key: string]: string;
}

export interface IosSaf {
  [key: string]: string;
}

export interface OpMini {
  [key: string]: string;
}

export interface Android {
  [key: string]: string;
}

export interface Bb {
  [key: string]: string;
}

export interface OpMob {
  [key: string]: string;
}

export interface AndChr {
  [key: string]: string;
}

export interface AndFf {
  [key: string]: string;
}

export interface IeMob {
  [key: string]: string;
}

export interface AndUc {
  [key: string]: string;
}

export interface Samsung {
  [key: string]: string;
}

export interface AndQq {
  [key: string]: string;
}

export interface Baidu {
  [key: string]: string;
}

export interface Kaios {
  [key: string]: string;
}

export interface Stats {
  ie: Ie;
  edge: Edge;
  firefox: Firefox;
  chrome: Chrome;
  safari: Safari;
  opera: Opera;
  ios_saf: IosSaf;
  op_mini: OpMini;
  android: Android;
  bb: Bb;
  op_mob: OpMob;
  and_chr: AndChr;
  and_ff: AndFf;
  ie_mob: IeMob;
  and_uc: AndUc;
  samsung: Samsung;
  and_qq: AndQq;
  baidu: Baidu;
  kaios: Kaios;
}

export type ExtendedCanIUseData = {
  description: string;
  spec: string;
  status: string;
  stats: Stats;
  bug_count: number;
  link_count: number;
  notes: string;
  notes_by_num: { [key: number]: string };
  ucprefix: boolean;
  parent: string;
  children?: any;
  keywords: string;
  date_published: string;
  ie_id: string;
  chrome_id: string;
  firefox_id: string;
  webkit_id: string;
  baseCategories: string[];

  title: string;
  path?: string;
  support: {
    chrome: Chrome;
    and_chr: AndChr;
    edge: Edge;
    firefox: Firefox;
    and_ff: AndFf;
    ie: Ie;
    opera: Opera;
    op_mob: OpMob;
    safari: Safari;
    ios_saf: IosSaf;
    samsung: Samsung;
    android: Android;
  };
  amountOfBrowsersWithData: number;
  mdnStatus: {
    experimental: boolean;
    standard_track: boolean;
    deprecated: boolean;
  };
  mdn_url: string;
};
