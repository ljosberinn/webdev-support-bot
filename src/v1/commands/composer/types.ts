export interface PackagistResponse {
  results: {
    name: string;
    description: string;
    url: string;
    repository: string;
    downloads: number;
    favers: number;
  }[];
  total: number;
  next?: string;
}

export interface ExtendedPackagistResponse {
  package: {
    name: string;
    description: string;
    time: Date;
    maintainers: {
      name: string;
      avatar_url: string;
    }[];
    versions: Versions;
    type: string;
    repository: string;
    github_stars: number;
    github_watchers: number;
    github_forks: number;
    github_open_issues: number;
    language: string;
    dependents: number;
    suggesters: number;
    downloads: Downloads;
    favers: number;
  };
}

export interface Versions {
  'dev-master': Version; // dev-master is only a pointer to an actual version we care about
  [otherVersions: string]: Version;
}

export interface Version {
  name: string;
  description: string;
  keywords: string[];
  homepage: string;
  version: string;
  version_normalized: string;
  license: string[];
  funding: boolean;
  time: Date;
  source: {
    type: string;
    url: string;
    reference: string;
  };
  require: { [key: string]: string };
  authors: {
    name: string;
    email: string;
  }[];
}

export interface Downloads {
  total: number;
  monthly: number;
  daily: number;
}
