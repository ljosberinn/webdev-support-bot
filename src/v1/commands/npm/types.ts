export interface Links {
  npm: string;
  homepage: string;
  repository: string;
  bugs: string;
}

export interface Author {
  name: string;
  email: string;
}

export interface Publisher {
  username: string;
  email: string;
}

export interface Maintainer {
  username: string;
  email: string;
}

export interface NPMResponse {
  name: string;
  scope: string;
  version: string;
  description: string;
  keywords: string[];
  date: Date;
  links: Links;
  author: Author;
  publisher: Publisher;
  maintainers: Maintainer[];
}
