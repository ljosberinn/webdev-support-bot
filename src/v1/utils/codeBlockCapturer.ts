const BACKTICKS = '```';

export type CodeBlockData = {
  code: string;
  language: string;
};

export function createCodeBlockCapturer(langs: string[] = []) {
  const langAlts = langs.join('|');
  return function* (str: string): Generator<CodeBlockData> {
    const langRegex = new RegExp(
      String.raw`${BACKTICKS}(?<language>${langAlts})\n(?<code>[\s\S]+?)\n${BACKTICKS}`,
      'gi'
    );
    const matches = str.matchAll(langRegex);
    for (const {
      groups: { language, code },
    } of matches) {
      yield {
        language: language || '',
        code: code || '',
      };
    }
  };
}
