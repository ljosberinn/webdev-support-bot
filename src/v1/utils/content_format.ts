import { Message } from 'discord.js';

const linebreakPattern = /\n/gim;

export const generateCleanContent = (msg: Message) =>
  msg.cleanContent.replace(linebreakPattern, ' ').toLowerCase();

export const stripMarkdownQuote = (msg: string) => msg.replace(/^> .+$/gm, '');
