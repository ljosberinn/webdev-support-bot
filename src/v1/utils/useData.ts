import * as Sentry from '@sentry/node';
import fetch, {
  HeaderInit,
  RequestInfo,
  RequestInit,
  Response,
} from 'node-fetch';

import {
  API_CACHE_ENTRIES_LIMIT,
  API_CACHE_EXPIRATION_IN_SECONDS,
  API_CACHE_REVALIDATION_WINDOW_IN_SECONDS,
} from '../env';
import { Cache } from './Cache';

const apiCache = new Cache({
  checkperiod: Number.parseInt(API_CACHE_REVALIDATION_WINDOW_IN_SECONDS, 10),
  maxKeys: Number.parseInt(API_CACHE_ENTRIES_LIMIT, 10),
  stdTTL: Number.parseInt(API_CACHE_EXPIRATION_IN_SECONDS, 10),
});

type ResponseTypes = 'json' | 'text';

type UnknownData<T> =
  | {
      error: true;
      json: null;
      text: null;
    }
  | {
      error: false;
      text: null;
      json: T;
    }
  | {
      error: false;
      json: null;
      text: string;
    };

type ResponseMapper<Data> = (response: Response) => Promise<Data>;
type FetchWithFormat<Format> = (
  url: RequestInfo,
  init?: RequestInit
) => Promise<Format>;

/**
 * Given a cache key and a function that maps the fetch response to a arbitrary
 * data structure, return a function when invoked, lower cases the cache key and
 * returns the cached formatted response when found.
 *
 * If the cache entry is _not_ found, delegate the call to an actual fetch implementation
 * to make the HTTP request. This response is then formatted with the function
 * above and saved into the cache. HTTP requests times/durations are also logged.
 */
const doFetch: <TParsedResponse>(
  cacheKey: string,
  mapper: ResponseMapper<TParsedResponse>
) => FetchWithFormat<TParsedResponse> = <TParsedResponse>(cacheKey, mapper) => {
  const casedCacheKey = cacheKey.toLowerCase();
  const cachedResponse = apiCache.get<TParsedResponse>(casedCacheKey);

  if (cachedResponse) {
    return () => cachedResponse;
  }

  return async (url, fetchOptions) => {
    Sentry.addBreadcrumb({
      category: 'query',
      data: typeof url === 'string' ? { url } : undefined,
      level: Sentry.Severity.Info,
      timestamp: Date.now(),
    });

    const timeLabel = `Time took for url=${encodeURIComponent(url.toString())}`;
    // eslint-disable-next-line no-console
    console.time(timeLabel);
    const response = await fetch(url, fetchOptions);
    // eslint-disable-next-line no-console
    console.timeEnd(timeLabel);
    const formattedResponse = await mapper(response);
    apiCache.set(casedCacheKey, formattedResponse);
    return formattedResponse;
  };
};

const responseMapper: <T>(
  type: ResponseTypes
) => (response: Response) => Promise<UnknownData<T>> = (
  type: string
) => async response => {
  if (!response.ok) {
    return {
      error: true,
      json: null,
      text: null,
    };
  }

  if (type === 'json') {
    const json = await response.json();
    return {
      error: false,
      json,
      text: null,
    };
  }

  if (type === 'text') {
    const text = await response.text();

    return {
      error: false,
      json: null,
      text,
    };
  }
};

const useData = <T>(
  url: string,
  type: ResponseTypes = 'json',
  headers: HeaderInit = {}
): Promise<UnknownData<T>> => {
  return doFetch(url, responseMapper<T>(type))(url, { headers });
};

export default useData;
