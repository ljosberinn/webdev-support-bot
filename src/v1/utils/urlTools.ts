import { Message } from 'discord.js';
import { HeadersInit } from 'node-fetch';

import { delayedMessageAutoDeletion } from './delayedMessageAutoDeletion';
import { Provider } from './discordTools';
import { noResults, invalidResponse } from './errors';
import useData from './useData';

const SEARCH_TERM = '%SEARCH%';
const TERM = '%TERM%';

type ProviderMap = {
  [key in Provider]: {
    search: string;
    color: number;
    createTitle: (term: string) => string;
    icon: string;
    help: string;
    direct?: string;
    getExtendedInfoUrl?: (pkg: string) => string;
  };
};

export const providers: ProviderMap = {
  bundlephobia: {
    color: 0xffffff,
    createTitle: (searchTerm: string) =>
      `Bundlephobia results for *${searchTerm}*`,
    direct: `https://bundlephobia.com/result?p=${TERM}`,
    getExtendedInfoUrl: (pkg: string) =>
      `https://bundlephobia.com/api/size?package=${pkg}&record=true`,
    help: '!bundlephobia @chakra-ui/core',
    icon: 'https://bundlephobia.com/android-chrome-192x192.png',
    search: `https://api.npms.io/v2/search/suggestions?q=${SEARCH_TERM}`,
  },
  caniuse: {
    color: 0xdb5600,
    createTitle: (searchTerm: string) => `CanIUse results for *${searchTerm}*`,
    direct: `https://caniuse.com/#feat=${TERM}`,
    getExtendedInfoUrl: (text: string) =>
      `https://caniuse.com/process/get_feat_data.php?type=support-data&feat=${text}`,
    help: '!caniuse IntersectionObserver',
    icon: 'https://caniuse.com/img/favicon-128.png',
    search: `https://caniuse.com/process/query.php?search=${SEARCH_TERM}`,
  },
  composer: {
    color: 0xf28d1a,
    createTitle: (searchTerm: string) => `Packagist results for ${searchTerm}`,
    direct: `https://packagist.org/packages/${TERM}`,
    getExtendedInfoUrl: (pkg: string) =>
      `https://packagist.org/packages/${pkg}.json`,
    help: '!composer sentry',
    icon: 'https://packagist.org/bundles/packagistweb/img/logo-small.png',
    search: `https://packagist.org/search.json?q=${SEARCH_TERM}`,
  },
  github: {
    color: 0x24292e,
    createTitle: (searchTerm: string) => `GitHub results for *${searchTerm}*`,
    direct: `https://github.com/${TERM}`,
    help: '!github react',
    icon:
      'https://github.githubassets.com/images/modules/logos_page/GitHub-Mark.png',
    search: `https://api.github.com/search/repositories?q=${SEARCH_TERM}`,
  },
  mdn: {
    color: 0x83d0f2,
    createTitle: (searchTerm: string) => `MDN results for *${searchTerm}*`,
    direct: `https://developer.mozilla.org${TERM}`,
    help: '!mdn localStorage',
    icon: 'https://avatars0.githubusercontent.com/u/7565578',
    search: `https://developer.mozilla.org/api/v1/search?q=${SEARCH_TERM}&locale=en-US`,
  },
  npm: {
    color: 0xfb3e44,
    createTitle: (searchTerm: string) => `NPM results for *${searchTerm}*`,
    help: '!npm react',
    icon: 'https://avatars0.githubusercontent.com/u/6078720',
    search: `https://www.npmjs.com/search/suggestions?q=${SEARCH_TERM}`,
  },
  php: {
    color: 0x8892bf,
    createTitle: (searchTerm: string) => `PHP.net results for *${searchTerm}*`,
    direct: `https://www.php.net/${TERM}`,
    help: '!php echo',
    icon: 'https://www.php.net/images/logos/php-logo.svg',
    search: `https://www.php.net/${SEARCH_TERM}`,
  },
};

export const HELP_KEYWORD = '--help';
export const FORMATTING_KEYWORD = '!formatting';
export const FORMATTING_KEYWORD_ALT = '!format';
export const CODE_KEYWORD = '!code';
export const VSCODE_KEYWORD = '!vscode';
export const JOB_POSTING_KEYWORD = '!post';
export const JQUERY_KEYWORD = '!jquery';
export const POINTS_KEYWORD = '!points';
export const LEADERBOARD_KEYWORD = '!leaderboard';
export const DECAY_KEYWORD = '!decay';
export const MODULE_KEYWORD = '!modules';
export const LOCKFILE_KEYWORD = '!lockfile';
export const FLEXBOX_KEYWORD = '!flexbox';

/**
 * dynamic regExp matching all possible Object.keys(providers) as keyword
 */
export const KEYWORD_REGEXP = new RegExp(
  '^!(%PLACEHOLDER%)\\s+'.replace(
    '%PLACEHOLDER%',
    Object.keys(providers)
      .reduce((carry, keyword) => [...carry, keyword], [])
      .join('|')
  ),
  'i'
);

export const getSearchUrl = (provider: Provider, search: string) => {
  if (providers[provider]) {
    return providers[provider].search.replace(SEARCH_TERM, encodeURI(search));
  }

  throw new Error(`provider not implemeted: ${provider}`);
};

export const buildDirectUrl = (provider: Provider, href: string) => {
  if (providers[provider]) {
    return providers[provider].direct.replace(TERM, href);
  }

  throw new Error(`provider not implemeted: ${provider}`);
};

export const getExtendedInfoUrl = (provider: Provider, term: string) => {
  if (providers[provider] && providers[provider].getExtendedInfoUrl) {
    return providers[provider].getExtendedInfoUrl(term);
  }

  throw new Error(
    `provider or provider.getExtendedInfoUrl not implemented at provider: ${provider}`
  );
};

interface GetDataParams {
  msg: Message;
  provider: Provider;
  searchTerm: string;
  sanitizeData?: (data: any) => Partial<any>;
  isInvalidData: (data: any) => boolean;
  headers?: HeadersInit;
}

export const getData = async <T>({
  msg,
  provider,
  searchTerm,
  sanitizeData,
  isInvalidData,
  headers,
}: GetDataParams): Promise<Partial<T>> => {
  const searchUrl = getSearchUrl(provider, searchTerm);
  const { error, json: data } = await useData<T>(searchUrl, 'json', headers);

  if (error) {
    await msg.reply(invalidResponse);
    return;
  }

  const sanitizedData = sanitizeData ? sanitizeData(data) : data;

  if (isInvalidData(sanitizedData)) {
    const sentMessage = await msg.reply(noResults(searchTerm));

    delayedMessageAutoDeletion(sentMessage);
    return;
  }

  return sanitizedData;
};
