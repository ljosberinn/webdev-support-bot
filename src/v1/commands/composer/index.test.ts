import { getChosenResult } from '../../utils/discordTools';
import * as errors from '../../utils/errors';
import { getData } from '../../utils/urlTools';
import useData from '../../utils/useData';
import { PackagistResponse } from './types';

import { buildComposerQueryHandler } from '.';

describe('handleComposerQuery', () => {
  const sendMock = jest.fn();
  const replyMock = jest.fn();
  const msg: any = {
    channel: { send: sendMock },
    delete: replyMock,
    reply: replyMock,
  };

  const fetch: jest.MockedFunction<typeof getData> = jest.fn();
  const fetchUse: jest.MockedFunction<typeof useData> = jest.fn();
  const choose: jest.MockedFunction<typeof getChosenResult> = jest.fn();

  afterEach(() => jest.resetAllMocks());

  test('does nothing when packagist request is invalid', async () => {
    fetch.mockResolvedValue(null);
    const handler = buildComposerQueryHandler(fetch);
    await handler(msg, 'search term');

    expect(fetch).toBeCalledWith({
      isInvalidData: expect.any(Function),
      msg,
      provider: 'composer',
      searchTerm: 'search term',
    });

    expect(msg.channel.send).not.toHaveBeenCalled();
  });

  test('returns when chosen result is not given', async () => {
    const packageRes: PackagistResponse = {
      results: [
        {
          description: 'Better than Laravel',
          downloads: 10000,
          favers: 10000,
          name: 'Symfony',
          repository: 'https://github.com/symfony/symfony',
          url: 'https://symfony.com',
        },
      ],
      total: 1,
    };

    fetch.mockResolvedValue(packageRes);
    // eslint-disable-next-line unicorn/no-useless-undefined
    choose.mockResolvedValue(undefined);
    msg.channel.send.mockResolvedValue(null);

    const handler = buildComposerQueryHandler(fetch, fetchUse, choose);
    await handler(msg, 'search term');

    expect(fetch).toBeCalledWith({
      isInvalidData: expect.any(Function),
      msg,
      provider: 'composer',
      searchTerm: 'search term',
    });

    expect(msg.channel.send).toHaveBeenCalledWith(
      expect.objectContaining({
        embed: expect.objectContaining({
          description: expect.stringContaining('Symfony'),
        }),
      })
    );
    expect(fetchUse).not.toBeCalled();
  });

  test('calls the use useData method and fails', async () => {
    const packageRes: PackagistResponse = {
      results: [
        {
          description: 'Better than Laravel',
          downloads: 10000,
          favers: 10000,
          name: 'Symfony',
          repository: 'https://github.com/symfony/symfony',
          url: 'https://symfony.com',
        },
      ],
      total: 1,
    };

    fetch.mockResolvedValue(packageRes);
    choose.mockResolvedValue({
      description: 'Better than Laravel',
      downloads: 10000,
      name: 'Symfony',
      repository: 'https://github.com/symfony/symfony',
      stars: 10000,
      url: 'https://symfony.com',
    });
    fetchUse.mockResolvedValue({
      error: true,
      json: null,
      text: null,
    });

    msg.channel.send.mockResolvedValue(null);

    const handler = buildComposerQueryHandler(fetch, fetchUse, choose);
    await handler(msg, 'search term');

    expect(fetch).toBeCalledWith({
      isInvalidData: expect.any(Function),
      msg,
      provider: 'composer',
      searchTerm: 'search term',
    });

    const payload = msg.channel.send.mock.calls[0][0];
    expect(payload).toMatchSnapshot();
    expect(msg.reply).toBeCalledWith(errors.invalidResponse);
  });

  test('calls the use useData method and succeeds', async () => {
    const editMock = jest.fn();
    msg.channel.send.mockResolvedValue({ edit: editMock });

    const packageRes: PackagistResponse = {
      results: [
        {
          description: 'Better than Laravel',
          downloads: 10000,
          favers: 10000,
          name: 'Symfony',
          repository: 'https://github.com/symfony/symfony',
          url: 'https://symfony.com',
        },
      ],
      total: 1,
    };

    fetch.mockResolvedValue(packageRes);
    choose.mockResolvedValue({
      description: 'Better than Laravel',
      downloads: 10000,
      name: 'Symfony',
      repository: 'https://github.com/symfony/symfony',
      stars: 10000,
      url: 'https://symfony.com',
    });
    fetchUse.mockResolvedValue({
      error: false,
      json: {
        package: {
          dependents: 10,
          description: 'Better than Laravel',
          downloads: {
            daily: 1234,
            monthly: 1234,
            total: 1000000,
          },
          favers: 1234,
          github_forks: 1000,
          github_open_issues: 0,
          github_stars: 10000,
          github_watchers: 10000,
          language: 'PHP',
          maintainers: [
            {
              avatar_url: 'http://avatar',
              name: 'A Name',
            },
          ],
          name: 'Symfony',
          repository: 'https://github.com/symfony/symfony',
          suggesters: 1000,
          time: new Date('2020-01-01'),
          type: 'Framework',
          versions: {
            '1.0.0': {
              authors: [{ name: 'Example ' }],
              description: 'master',
              homepage: 'https://github.com/symfony/symfony',
              keywords: ['PHP'],
              license: ['MIT'],
              name: 'master',
              require: ['php', '7'],
              time: new Date('2020-01-01'),
              version: '1.0.0',
              version_normalized: '1.0.0',
            },
            'dev-master': {
              authors: [{ name: 'Example ' }],
              description: 'master',
              homepage: 'https://github.com/symfony/symfony',
              keywords: ['PHP'],
              license: ['MIT'],
              name: 'master',
              require: ['php', '7'],
              time: new Date('2020-01-01'),
              version: '1.0.0',
              version_normalized: '1.0.0',
            },
          },
        },
      },
      text: null,
    });

    const handler = buildComposerQueryHandler(
      fetch,
      fetchUse,
      choose,
      () => '5 months'
    );
    await handler(msg, 'search term');

    expect(fetch).toBeCalledWith({
      isInvalidData: expect.any(Function),
      msg,
      provider: 'composer',
      searchTerm: 'search term',
    });

    const payload = msg.channel.send.mock.calls[0][0];
    expect(payload).toMatchSnapshot();

    expect(msg.channel.send).toHaveBeenCalledWith(
      expect.objectContaining({
        embed: expect.objectContaining({
          description: expect.stringContaining('Symfony'),
        }),
      })
    );

    const editPayload = sendMock.mock.calls[0][0];
    expect(editPayload).toMatchSnapshot();
    expect(editMock).toBeCalled();
    expect(editMock.mock.calls[0][0]).toMatchSnapshot();
  });
});
