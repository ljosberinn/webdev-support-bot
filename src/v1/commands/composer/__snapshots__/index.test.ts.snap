// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleComposerQuery calls the use useData method and fails 1`] = `
Object {
  "embed": Object {
    "author": null,
    "color": 15895834,
    "description": "1. [**Symfony** - *Better than Laravel*](https://symfony.com)

:bulb: *react with a number (:one:, :two:, ...) to filter your result*
:neutral_face: *react with \`❌\` to delete*
:point_up: *supports \`!mdn\`, \`!github\`, \`!caniuse\`, \`!npm\`, \`!composer\`, \`!bundlephobia\`, and \`!php\`*
:gear: *issues? feature requests? head over to [github](https://github.com/ljosberinn/webdev-support-bot)*",
    "fields": Array [],
    "footer": Object {
      "iconURL": "https://packagist.org/bundles/packagistweb/img/logo-small.png",
      "text": "1 packages found",
    },
    "title": "Packagist results for search term",
    "url": "https://packagist.org/?query=search%20term",
  },
}
`;

exports[`handleComposerQuery calls the use useData method and succeeds 1`] = `
Object {
  "embed": Object {
    "author": null,
    "color": 15895834,
    "description": "1. [**Symfony** - *Better than Laravel*](https://symfony.com)

:bulb: *react with a number (:one:, :two:, ...) to filter your result*
:neutral_face: *react with \`❌\` to delete*
:point_up: *supports \`!mdn\`, \`!github\`, \`!caniuse\`, \`!npm\`, \`!composer\`, \`!bundlephobia\`, and \`!php\`*
:gear: *issues? feature requests? head over to [github](https://github.com/ljosberinn/webdev-support-bot)*",
    "fields": Array [],
    "footer": Object {
      "iconURL": "https://packagist.org/bundles/packagistweb/img/logo-small.png",
      "text": "1 packages found",
    },
    "title": "Packagist results for search term",
    "url": "https://packagist.org/?query=search%20term",
  },
}
`;

exports[`handleComposerQuery calls the use useData method and succeeds 2`] = `
Object {
  "embed": Object {
    "author": null,
    "color": 15895834,
    "description": "1. [**Symfony** - *Better than Laravel*](https://symfony.com)

:bulb: *react with a number (:one:, :two:, ...) to filter your result*
:neutral_face: *react with \`❌\` to delete*
:point_up: *supports \`!mdn\`, \`!github\`, \`!caniuse\`, \`!npm\`, \`!composer\`, \`!bundlephobia\`, and \`!php\`*
:gear: *issues? feature requests? head over to [github](https://github.com/ljosberinn/webdev-support-bot)*",
    "fields": Array [],
    "footer": Object {
      "iconURL": "https://packagist.org/bundles/packagistweb/img/logo-small.png",
      "text": "1 packages found",
    },
    "title": "Packagist results for search term",
    "url": "https://packagist.org/?query=search%20term",
  },
}
`;

exports[`handleComposerQuery calls the use useData method and succeeds 3`] = `
Object {
  "embed": Object {
    "author": Object {
      "icon_url": "http://avatar",
      "name": "A Name",
    },
    "color": 15895834,
    "description": "Better than Laravel",
    "fields": Array [
      Object {
        "inline": false,
        "name": "add to your project",
        "value": "\`\`\`bash
composer require master
\`\`\`",
      },
      Object {
        "inline": false,
        "name": "PHP keywords",
        "value": "[PHP](https://packagist.org/search/?tags=PHP)",
      },
      Object {
        "inline": true,
        "name": ":chains: dependencies",
        "value": "2",
      },
      Object {
        "inline": true,
        "name": "MIT license",
        "value": "[MIT](https://choosealicense.com/licenses/mit)",
      },
      Object {
        "inline": true,
        "name": "​",
        "value": "​",
      },
      Object {
        "inline": true,
        "name": ":globe_with_meridians: homepage",
        "value": "[github.com/symfony/symfony](https://github.com/symfony/symfony)",
      },
      Object {
        "inline": true,
        "name": "​",
        "value": "​",
      },
      Object {
        "inline": true,
        "name": ":writing_hand: author",
        "value": "Example ",
      },
    ],
    "footer": Object {
      "iconURL": "https://packagist.org/bundles/packagistweb/img/logo-small.png",
      "text": "Downloads: 1,000,000 total | 1,234 monthly | 1,234 daily
last updated 5 months ago",
    },
    "title": "Symfony *(1.0.0)*",
    "url": "https://packagist.org/packages/Symfony",
  },
}
`;
